modules/mcaptcha.nix

```diff
@@ -0,0 +1,279 @@
+{
+  config,
+  lib,
+  options,
+  pkgs,
+  ...
+}:
+with builtins;
+with lib; let
+  cfg = config.services.mcaptcha;
+
+  settingsFormat = pkgs.formats.toml {};
+  filteredSettings = lib.attrsets.filterAttrsRecursive (_path: value: value != null) cfg.settings;
+  configFile = settingsFormat.generate "mcaptcha.config.toml" filteredSettings;
+in {
+  options.services.mcaptcha.enable = mkEnableOption "Enable mCaptcha server.";
+  options.services.mcaptcha.package = mkPackageOption pkgs "mcaptcha" {};
+
+  options.services.mcaptcha.settings = lib.mkOption {
+    type = lib.types.submodule {
+      freeformType = settingsFormat.type;
+
+      options.database.name = mkOption {
+        type = types.str;
+        description = "Applies both when `database.createLocally` is set and not.";
+        default = "mcaptcha";
+      };
+
+      options.database.username = mkOption {
+        type = types.nullOr types.str;
+        description = "Ignored when `database.createLocally`.";
+        example = "mcaptcha";
+        default = null;
+      };
+
+      options.database.hostname = mkOption {
+        type = types.nullOr types.str;
+        description = "Ignored when `database.createLocally`.";
+        example = "localhost";
+        default = null;
+      };
+
+      options.database.port = mkOption {
+        type = types.nullOr types.port;
+        description = "Ignored when `database.createLocally`.";
+        example = 5432;
+        default = null;
+      };
+
+      options.server.port = mkOption {
+        type = types.port;
+        description = "Web server port.";
+        default = 7000;
+      };
+
+      options.server.domain = mkOption {
+        type = types.str;
+        description = "Web server host.";
+        default = "localhost";
+        example = "example.com";
+      };
+
+      options.server.ip = mkOption {
+        type = types.str;
+        description = "Web server addresses to bind to.";
+        default = "127.0.0.1";
+        example = "0.0.0.0";
+      };
+    };
+
+    description = ''
+      Extra settings. Best sources of documentation for settings seem to be
+      https://github.com/mCaptcha/mCaptcha/blob/master/config/default.toml
+      https://github.com/mCaptcha/mCaptcha/blob/master/docs/CONFIGURATION.md
+    '';
+  };
+
+  options.services.mcaptcha.user = mkOption {
```

Comment:
Again, repetition of `options.services.mcaptcha` are too verbose for my taste. I'd prefer:

```nix
options.services.mcaptcha = {
  user = mkOption { /* ... */ };
  /* ... */
}
```